frontend/src/globals/unity-indexeddb-worker.ts

```diff
@@ -0,0 +1,359 @@
+/**
+ *  Copyright (C) 2022 3D Repo Ltd
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * This entire script/file is run inside a WebWorker to service requests from
+ * an instance of IndexedDbCache. IndexedDbCache posts messages to this worker
+ * which are received through the locally scoped onmessage function. Messages
+ * are sent back with self.postMessage, where self is the WorkerGlobalScope
+ * the worker runs in. IndexedDbCacheWorker is a convenience class that
+ * encompasses the actual IndexedDb API usage. An instance is created for use
+ * by this worker.
+ */
+
+class IndexedDbCacheWorker {
+	constructor() {
+		this.objectStoreName = '3DRepoCache';
+		this.memoryCache = {};
+		this.numReadTransactions = 0;
+		this.createIndexedDbCache();
+
+		// Check at low frequency if there is anything going on, and if not try
+		// to commit the memory cache.
+		// This is because writes are typically started after all reads are done
+		// but if no reads happen for a while then no writes will be dispatched!
+		setInterval(() => {
+			this.commitMemoryCache();
+		},
+		2000);
+	}
+
+	/** Reference to the open database. */
+	db: any;
+
+	/** Name of the object store for the cache. This is set once when the db is
+	 * created. */
+	objectStoreName: string;
+
+	/** A dictionary of records that are waiting to be committed. */
+	memoryCache: any;
+
+	/** How many read transactions are currently created. We always want read
+	 * transactions to take priority over writes, so when this is non-zero
+	 * the write chain is interrupted. */
+	numReadTransactions: number;
+
+	/** Is there a running readwrite transaction? If so don't start another one
+	 * or we may commit records multiple times (wasting time). */
+	committing: boolean;
+
+	/** The local index is used to keep track of which keys are in the object
+	 * store(s). This is initialised when the database is opened and updated
+	 * when set requests are made. It is used to perform early rejects on get
+	 * requests we know will fail (because there is no key in the store),
+	 * avoiding expensive database transactions. */
+	index: any;
+
+	createIndexedDbCache() {
+		const request = indexedDB.open('3DRepoCacheDb', 1);
+
+		request.onerror = () => {
```

Comment:
could we make these camel case or is the naming convention imposed on us by something else?

Reply:
It is the `IDBOpenDBRequest` [API](https://developer.mozilla.org/en-US/docs/Web/API/IDBOpenDBRequest) unfortunately!